cmd/deploy/proxy.go

```diff
@@ -0,0 +1,182 @@
+package deploy
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+)
+
+type proxyConfig struct {
+	port         int
+	certificates []tls.Certificate
+	token        string
+}
+
+type proxy struct {
+	s           *http.Server
+	proxyConfig proxyConfig
+}
+
+func newProxy(proxyConfig proxyConfig) *proxy {
+	return &proxy{
+		proxyConfig: proxyConfig,
+	}
+}
+
+// Start starts the proxy server
+func (p *proxy) Start(ctx context.Context, name string, clusterConfig *rest.Config) error {
+	handler, err := p.createProxyHandler(ctx, name, clusterConfig)
+	if err != nil {
+		return err
+	}
+
+	p.createAndStartProxyServer(ctx, handler, p.proxyConfig.port, p.proxyConfig.certificates)
+
+	return nil
+}
+
+func (p *proxy) createAndStartProxyServer(ctx context.Context, handler http.Handler, port int, certs []tls.Certificate) {
+	s := &http.Server{
+		Addr:         fmt.Sprintf(":%d", port),
+		Handler:      handler,
+		ReadTimeout:  3600 * time.Second,
+		WriteTimeout: 3600 * time.Second,
+		TLSConfig: &tls.Config{
+			Certificates: certs,
+		},
+	}
+
+	go func(s *http.Server) {
+		log.Debugf("start server on %d", port)
+		// Path to cert and key files are empty because cert is provisioned on the tls config struct
+		if err := s.ListenAndServeTLS("", ""); err != nil && err != http.ErrServerClosed {
+			log.Fatalf("could not start proxy server: %s", err)
+		}
+	}(s)
+
+	p.s = s
+}
+
+func (p *proxy) createProxyHandler(ctx context.Context, name string, clusterConfig *rest.Config) (http.Handler, error) {
+	trans, err := rest.TransportFor(clusterConfig)
+	if err != nil {
+		log.Errorf("could not get transport from config: %s", err)
+		return nil, err
+	}
+
+	handler := http.DefaultServeMux
+
+	proxy := httputil.NewSingleHostReverseProxy(&url.URL{
+		Host:   strings.TrimPrefix(clusterConfig.Host, "https://"),
+		Scheme: "https",
+	})
+	proxy.Transport = trans
+
+	log.Debugf("forwarding host: %s", clusterConfig.Host)
+
+	handler.HandleFunc("/", func(rw http.ResponseWriter, r *http.Request) {
+		token := r.Header.Get("Authorization")
+		expectedToken := fmt.Sprintf("Bearer %s", p.proxyConfig.token)
+		// Validate token with the generated for the local kubeconfig file
+		if token != expectedToken {
+			rw.WriteHeader(401)
+			return
+		}
+
+		// Set the right bearer token based on the original kubeconfig
+		r.Header.Set("Authorization", fmt.Sprintf("Bearer %s", clusterConfig.BearerToken))
+
+		// Modify all resources updated or created to include the label.
+		if r.Method == "PUT" || r.Method == "POST" {
+			b, err := io.ReadAll(r.Body)
+			if err != nil {
+				log.Errorf("could not read the request body: %s", err)
+				rw.WriteHeader(500)
```

Comment:
What do we want to do in case of some errors processing the body request? For now I'm failing with a 500 error but this would prevent any kubernetes command to be executed (just in the scope of the `okteto deploy` command. User's kubeconfig won't be modified)

Reply:
Yeap, I think it is the best we can do